subsys/shell/shell_telnet.c

```diff
@@ -0,0 +1,493 @@
+
+#define TELNET_MIN_MSG 2
+
+/* Basic TELNET implmentation. */
+
+static void telnet_end_client_connection(void)
+{
+	struct net_pkt *pkt;
+
+	net_context_put(sh_telnet->client_ctx);
+	sh_telnet->client_ctx = NULL;
+	sh_telnet->output_lock = false;
+
+	k_timer_stop(&sh_telnet->send_timer);
+
+	/* Flush the RX FIFO */
+	while ((pkt = k_fifo_get(&sh_telnet->rx_fifo, K_NO_WAIT)) != NULL) {
+		net_pkt_unref(pkt);
+	}
+}
+
+static void telnet_sent_cb(struct net_context *client,
+			   int status, void *user_data)
+{
+	if (status < 0) {
+		telnet_end_client_connection();
+		LOG_ERR("Could not send packet %d", status);
+	}
+}
+
+#ifdef CONFIG_SHELL_BACKEND_TELNET_SUPPORT_COMMAND
+
+static void telnet_command_send_reply(u8_t *msg, u16_t len)
+{
+	int err;
+
+	err = net_context_send(sh_telnet->client_ctx, msg, len, telnet_sent_cb,
+			       K_FOREVER, NULL);
+	if (err < 0) {
+		NET_ERR("Failed to send command %d, shutting down", err);
+		telnet_end_client_connection();
+	}
+}
+
+static void telnet_reply_ay_command(void)
+{
+	static const char alive[] = "Zephyr at your service\r\n";
+
+	telnet_command_send_reply((u8_t *)alive, strlen(alive));
+}
+
+static void telnet_reply_do_command(struct telnet_simple_command *cmd)
+{
+	switch (cmd->opt) {
+	case NVT_OPT_SUPR_GA:
+		cmd->op = NVT_CMD_WILL;
+		break;
+	default:
+		cmd->op = NVT_CMD_WONT;
+		break;
+	}
+
+	telnet_command_send_reply((u8_t *)cmd,
+				  sizeof(struct telnet_simple_command));
+}
+
+static void telnet_reply_command(struct telnet_simple_command *cmd)
+{
+	if (!cmd->iac) {
+		return;
+	}
+
+	switch (cmd->op) {
+	case NVT_CMD_AO:
+		/* OK, no output then */
+		sh_telnet->output_lock = true;
+		sh_telnet->line_out.len = 0;
+		k_timer_stop(&sh_telnet->send_timer);
+		break;
+	case NVT_CMD_AYT:
+		telnet_reply_ay_command();
+		break;
+	case NVT_CMD_DO:
+		telnet_reply_do_command(cmd);
+		break;
+	default:
+		LOG_DBG("Operation %u not handled", cmd->op);
+		break;
+	}
+}
+
+#endif /* CONFIG_SHELL_BACKEND_TELNET_SUPPORT_COMMAND */
+
+static int telnet_send(void)
+{
+	int err;
+
+	if (sh_telnet->line_out.len == 0) {
+		return 0;
+	}
+
+	err = net_context_send(sh_telnet->client_ctx, sh_telnet->line_out.buf,
+			       sh_telnet->line_out.len, telnet_sent_cb,
+			       K_FOREVER, NULL);
+	if (err < 0) {
+		NET_ERR("Failed to send %d, shutting down", err);
+		telnet_end_client_connection();
+		return err;
+	}
+
+	/* We reinitialize the line buffer */
+	sh_telnet->line_out.len = 0;
+
+	return 0;
+}
+
+static void telnet_send_prematurely(struct k_timer *timer)
+{
+	(void)telnet_send();
+}
+
+static inline bool telnet_handle_command(struct net_pkt *pkt)
+{
+	/* Commands are two or three bytes. */
+	NET_PKT_DATA_ACCESS_CONTIGUOUS_DEFINE(cmd_access, u16_t);
+	struct telnet_simple_command *cmd;
+
+	cmd = (struct telnet_simple_command *)net_pkt_get_data(pkt,
+							       &cmd_access);
+	if (!cmd || cmd->iac != NVT_CMD_IAC) {
+		return false;
+	}
+
+#ifdef CONFIG_SHELL_BACKEND_TELNET_SUPPORT_COMMAND
```

Comment:
use IS_ENABLED() and you won't need `#ifdef` around function implementation as well.